node/activation_service.go

```diff
@@ -0,0 +1,712 @@
+	listenAddress := app.Config.BaseConfig.NodeServiceAddress
+	logger := app.addLogger(NodeServiceClientLogger, lg).Zap()
+	cfg := &nodeclient.Config{
+		RetryWaitMin: time.Millisecond * 500,
+		RetryWaitMax: time.Second,
+		RetryMax:     10,
+	}
+	var err error
+	nodeServiceClient, err = nodeclient.NewNodeServiceClient(listenAddress, logger, cfg)
+	if err != nil {
+		return fmt.Errorf("creating node service client: %w", err)
+	}
+
+	poetDb, err := activation.NewPoetDb(
+		app.db,
+		app.addLogger(PoetDbLogger, lg).Zap(),
+		activation.WithCacheSize(app.Config.POET.PoetProofsCache),
+		activation.WithRemotePoetStorer(nodeServiceClient),
+	)
+	if err != nil {
+		return fmt.Errorf("creating poet db: %w", err)
+	}
+	postStates := activation.NewPostStates(app.addLogger(PostLogger, lg).Zap())
+
+	app.idStates = identity.NewIdentityStateStorage(app.localDB, app.log.Zap())
+
+	opts := []activation.PostVerifierOpt{
+		activation.WithVerifyingOpts(app.Config.SMESHING.VerifyingOpts),
+		activation.WithAutoscaling(postStates),
+	}
+	for _, sig := range app.signers {
+		opts = append(opts, activation.WithPrioritizedID(sig.NodeID()))
+	}
+
+	verifier, err := activation.NewPostVerifier(
+		app.Config.POST,
+		app.addLogger(NipostValidatorLogger, lg).Zap(),
+		opts...,
+	)
+	if err != nil {
+		return fmt.Errorf("creating post verifier: %w", err)
+	}
+	app.postVerifier = verifier
+
+	validator := activation.NewValidator(
+		app.db,
+		poetDb,
+		app.Config.POST,
+		app.Config.SMESHING.Opts.Scrypt,
+		app.postVerifier,
+	)
+	app.validator = validator
+
+	goldenATXID := types.ATXID(app.Config.Genesis.GoldenATX())
+	if goldenATXID == types.EmptyATXID {
+		return errors.New("invalid golden atx id")
+	}
+
+	app.edVerifier = signing.NewEdVerifier(
+		signing.WithVerifierPrefix(app.Config.Genesis.GenesisID().Bytes()),
+	)
+
+	vrfVerifier := signing.NewVRFVerifier()
+	var beaconProtocol *beacon.ProtocolDriver
+
+	trtlCfg := app.Config.Tortoise
+	trtlCfg.LayerSize = layerSize
+	if trtlCfg.BadBeaconVoteDelayLayers == 0 {
+		trtlCfg.BadBeaconVoteDelayLayers = app.Config.LayersPerEpoch
+	}
+	trtlopts := []tortoise.Opt{
+		tortoise.WithLogger(app.addLogger(TrtlLogger, lg).Zap()),
+		tortoise.WithConfig(trtlCfg),
+	}
+	if trtlCfg.EnableTracer {
+		app.log.With().Info("tortoise will trace execution")
+		trtlopts = append(trtlopts, tortoise.WithTracer())
+	}
+	app.log.Info("initializing tortoise")
+	start := time.Now()
+	trtl, err := tortoise.Recover(
+		ctx,
+		app.db,
+		app.atxsdata,
+		app.clock.CurrentLayer(), trtlopts...,
+	)
+	if err != nil {
+		return fmt.Errorf("can't recover tortoise state: %w", err)
+	}
+	app.log.With().Info("tortoise initialized", log.Duration("duration", time.Since(start)))
+	if nodeServiceClient == nil {
+		app.eg.Go(func() error {
+			for rst := range beaconProtocol.Results() {
+				events.EmitBeacon(rst.Epoch, rst.Beacon)
+				trtl.OnBeacon(rst.Epoch, rst.Beacon)
+			}
+			app.log.Debug("beacon results watcher exited")
+			return nil
+		})
+	}
```

Comment:
This is a dead code because `nodeServiceClient != nil` always, isn't it?

Reply:
You are right! I overlooked it.